optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,397 @@
+    search_space: Dict[str, BaseDistribution],
+    rng: np.random.RandomState,
+    swapping_prob: float,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> Dict:
+
+    while True:
+        parents = _selection(crossover_name, study, parent_population, rng, dominates)
+        child = {}
+        transes = []
+        distributions = []
+        param_names = []
+        parents_not_categorical_params: List[List[np.float64]] = [[] for _ in range(len(parents))]
+        for param_name in search_space.keys():
+            param_distribution = search_space[param_name]
+            parents_param = [p.params[param_name] for p in parents]
+
+            # categorical data operates on uniform crossover
+            if isinstance(search_space[param_name], CategoricalDistribution):
+                param = _swap(parents_param[0], parents_param[-1], rng.rand(), swapping_prob)
+                child[param_name] = param
+                continue
+
+            trans = _SearchSpaceTransform({param_name: param_distribution})
+            transes.append(trans)
+            distributions.append(param_distribution)
+            param_names.append(param_name)
+            for parent_index, trial in enumerate(parents):
+                param = trans.transform({param_name: trial.params[param_name]})[0]
+                parents_not_categorical_params[parent_index].append(param)
+
+        xs = np.array(parents_not_categorical_params)
+
+        if crossover_name == "uniform":
+            params_array = _uniform(xs, rng, swapping_prob)
+        elif crossover_name == "blxalpha":
+            alpha = 0.5
+            params_array = _blxalpha(xs, rng, alpha)
+        elif crossover_name == "sbx":
+            if len(study.directions) == 1:
+                eta = 2
+            else:
+                eta = 20
+            params_array = _sbx(xs, rng, distributions, eta)
+        elif crossover_name == "vsbx":
+            if len(study.directions) == 1:
+                eta = 2
+            else:
+                eta = 20
+            params_array = _vsbx(xs, rng, eta)
+        elif crossover_name == "undx":
+            sigma_xi = 0.5
+            sigma_eta = 0.35 / np.sqrt(len(xs[0]))
+            params_array = _undx(xs, rng, sigma_xi, sigma_eta)
+        elif crossover_name == "undxm":
+            _m = 2
+            _n = len(xs[0])
+            sigma_xi = 1 / np.sqrt(_m)
+            sigma_eta = (
+                0.35
+                * np.sqrt(_m + 1)
+                * np.sqrt(3)
+                / np.sqrt(_n - _m)
+                / np.sqrt(_m + 2)
+                / np.sqrt(2)
+            )
+            params_array = _undxm(xs, rng, sigma_xi, sigma_eta)
+        elif crossover_name == "spx":
+            epsilon = np.sqrt(len(xs[0]) + 2)
+            params_array = _spx(xs, rng, epsilon)
+        else:
+            raise ValueError(f"{crossover_name} is not exist in optuna.")
```

Comment:
A ditto but we can assert that this code is never reached.
```suggestion
            assert False
```